### DISCUSSION TITLE
how to connect v2ray client in ubuntu server?

### DISCUSSION THREAD

**Discussion by Faarhad** [v2fly/v2ray-core#2474] [Q&A]
Hi everybody, I've downloaded and extracted v2ray on a Ubuntu server VM.
My aim is to pass the Ubuntu server traffic through the v2ray so it cloud bypass censorship.
how to run the v2ray client in the Ubuntu server without GUI? is there any help or documentation for this?
Thank you, guys.

**Comment by mydogshitgold** [comment]
Do you have ssh access to the server in question? As mentioned in the issues before, you can use systemd. This will require you to write a `.service' file and enable it so that the server can run it automatically on startup.

A configuration json file is also necessary for the v2ray. If you use GUI clients to connect to v2ray endpoints, it may be worthwhile to attempt exporting the json configuration from said clients, as the majority of them offer this functionality. There is also a tool called [vmess2json](https://github.com/boypt/vmess2json) that can convert vmess links to json.

**Comment by Faarhad** [comment]
Thank you for your reply @mydogshitgold,
yes, I exported my config.json and I have ssh access, I also can run v2ray and config file using this command:
`v2ray run -c /home/v2ray/config.json` and it returns  
[Warning] V2Ray 5.3.0 started
I'm sorry I'm a newbie and couldn't start the v2ray client as a service, would you please help me to config my Ubuntu server to run the v2ray client as a service? 
Thank for your help Sir.

**Reply by Faarhad** [reply]
Also in this help doc, it was only mentioned using the v2ray client in Ubuntu GUI 
https://umaint.github.io/2019/01/04/v2ray-ubuntu/
not in an environment without GUI like ubuntu server.

**Reply by mydogshitgold** [reply]
If you intend to run v2ray as a systemd service on a Linux distribution, you can follow the steps below to set it up:

1. Create a new systemd service file for v2ray at path `/etc/systemd/system/v2ray.service` with following content:

```ini
[Unit]
Description=V2Ray Service
After=network.target nss-lookup.target

[Service]
DynamicUser=yes
CapabilityBoundingSet=CAP_NET_ADMIN CAP_NET_BIND_SERVICE
AmbientCapabilities=CAP_NET_ADMIN CAP_NET_BIND_SERVICE
NoNewPrivileges=true
ExecStart=/usr/bin/v2ray run -config /etc/v2ray/config.json
Restart=on-failure
RestartPreventExitStatus=23

[Install]
WantedBy=multi-user.target
```
2. Reload the systemd daemon to ensure that the new service file is recognized: 

```sh
sudo systemctl daemon-reload
```
3. Create `/etc/v2ray/` directory and copy your `config.json` to it:
```
sudo mkdir -p /etc/v2ray
sudo cp /home/v2ray/config.json /etc/v2ray/config.json
```
4. Enable and start the v2ray service:
```sh
sudo systemctl enable --now v2ray.service
```
5. Check the status of the service to verify that it is running:
```
sudo systemctl status v2ray
```
the output should contain a line `Active: active (running)`

**Reply by mydogshitgold** [reply]
The basic operations of systemd are essential knowledge for effective server management, and you would benefit from familiarising yourself with them. To delve deeper into this topic, read the [wiki and manual pages from here](https://wiki.archlinux.org/title/Systemd). And you should edit the `config.json` file to suit your needs. I suggest reading the [v2ray documentation](https://www.v2fly.org/config/overview.html) (the content is mostly in Mandarin, you could use a [translatior](https://www.deepl.com/translator)) to get a better understanding of the process. As a side note, if you intend to use v2ray on a server, it would be wise to use a [transparent proxy](https://guide.v2fly.org/app/tproxy.html#%E5%BC%80%E6%9C%BA%E8%87%AA%E5%8A%A8%E8%BF%90%E8%A1%8C%E9%80%8F%E6%98%8E%E4%BB%A3%E7%90%86%E8%A7%84%E5%88%99), as many programs do not support explicit configuration of the socks/http proxy, making it easier to administer.

**Reply by Faarhad** [reply]
Thank you, Sir, I know I can use a transparent proxy, but the problem is the server is behind the censorship. I also export my own config.json just the same as the one I'm using for my personal use (in my phone using v2rayNG)
I created a service due to your help, but the service is not Active as a common service with the output below:
`
sudo systemctl status v2ray
● v2ray.service - V2Ray Service
     Loaded: loaded (/etc/systemd/system/v2ray.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Fri 2023-04-21 18:37:01 UTC; 7min ago
    Process: 939 ExecStart=/usr/bin/v2ray run -config /etc/v2ray/config.json (code=exited, status=203/EXEC)
   Main PID: 939 (code=exited, status=203/EXEC)

Apr 21 18:37:01 harbor systemd[1]: v2ray.service: Scheduled restart job, restart counter is at 5.
Apr 21 18:37:01 harbor systemd[1]: Stopped V2Ray Service.
Apr 21 18:37:01 harbor systemd[1]: v2ray.service: Start request repeated too quickly.
Apr 21 18:37:01 harbor systemd[1]: v2ray.service: Failed with result 'exit-code'.
Apr 21 18:37:01 harbor systemd[1]: Failed to start V2Ray Service.
`

**Reply by mydogshitgold** [reply]
what is the output of `journalctl -u v2ray`?

**Reply by Faarhad** [reply]
this is the output:

`Apr 21 18:34:22 harbor systemd[66509]: v2ray.service: Failed to execute command: No such file or directory
Apr 21 18:34:22 harbor systemd[66509]: v2ray.service: Failed at step EXEC spawning /usr/bin/v2ray: No such file or directory
Apr 21 18:34:22 harbor systemd[1]: v2ray.service: Main process exited, code=exited, status=203/EXEC
Apr 21 18:34:22 harbor systemd[1]: v2ray.service: Failed with result 'exit-code'.
Apr 21 18:34:22 harbor systemd[1]: v2ray.service: Scheduled restart job, restart counter is at 5.
Apr 21 18:34:22 harbor systemd[1]: Stopped V2Ray Service.
Apr 21 18:34:22 harbor systemd[1]: v2ray.service: Start request repeated too quickly.
Apr 21 18:34:22 harbor systemd[1]: v2ray.service: Failed with result 'exit-code'.
Apr 21 18:34:22 harbor systemd[1]: Failed to start V2Ray Service.
-- Reboot --
`

**Reply by mydogshitgold** [reply]
Where did you put the v2ray binary? Change the first part of `ExecStart` to the absolute path to that binary.

Then run `sudo systemctl daemon-reload && sudo systemctl restart v2ray`

**Reply by Faarhad** [reply]
Sir, I do the same, I've changed the v2ray binary path by a command like:
`mv v2ray /usr/bin`
and I do deamon reload and restart and etc.
but it remains "failed"
`Apr 21 18:34:22 harbor systemd[66509]: v2ray.service: Failed to execute the command: No such file or directory
Apr 21 18:34:22 harbor systemd[66509]: v2ray.service: Failed at step EXEC spawning /usr/bin/v2ray: No such file or directory
Apr 21 18:34:22 harbor systemd[1]: v2ray.service: Main process exited, code=exited, status=203/EXEC
Apr 21 18:34:22 harbor systemd[1]: v2ray.service: Failed with result 'exit-code'.
Apr 21 18:34:22 harbor systemd[1]: v2ray.service: Scheduled restart job, restart counter is at 5.
Apr 21 18:34:22 harbor systemd[1]: Stopped V2Ray Service.
Apr 21 18:34:22 harbor systemd[1]: v2ray.service: Start request repeated too quickly.
Apr 21 18:34:22 harbor systemd[1]: v2ray.service: Failed with result 'exit-code'.
Apr 21 18:34:22 harbor systemd[1]: Failed to start V2Ray Service.
-- Reboot --
Apr 21 18:37:00 harbor systemd[1]: Started V2Ray Service.
Apr 21 18:37:00 harbor systemd[887]: v2ray.service: Failed to execute command: No such file or directory
`
I'm sorry for my lack of knowledge to investigate the problem.

**Reply by Faarhad** [reply]
It may also help:
```
root@harbor:/etc/v2ray# /usr/bin/v2ray run -config /etc/v2ray/config.json
V2Ray 5.4.1 (V2Fly, a community-driven edition of V2Ray.) Custom (go1.20.2 linux/amd64)
A unified platform for anti-censorship.
Failed to start: main/commands: failed to load config: [/etc/v2ray/config.json] > infra/conf/rule: invalid field rule > infra/conf/rule: failed to parse domain rule: geosite:category-ads-all > infra/conf/rule: failed to load geosite: category-ads-all > infra/conf/geodata/standard: failed to open file: geosite.dat > open /usr/bin/geosite.dat: no such file or directory
root@harbor:/etc/v2ray#

```

**Reply by mydogshitgold** [reply]
You should not move `v2ray` into `/usr/bin`, as the binary downloaded from the release depends on the data files in the same directory. Move it back to the original path and change the `ExecStart` value.

**Reply by Faarhad** [reply]
Thank you for your help Sir, it's now Active.
you did me a great favor.

**Reply by mydogshitgold** [reply]
my pleasure.

**Comment by Faarhad** [comment]
@mydogshitgold 
Sir I have another question, I promise it's the last one =))
would you please help me how to bypass all my Ubuntu server through v2ray service we've created?
according to [this ](https://umaint.github.io/2019/01/04/v2ray-ubuntu/) it just configures to use a web browser like Firefox to use the proxy, I need to pass all traffic through it.

**Reply by mydogshitgold** [reply]
This sounds cryptic to me, can you state more clearly?

**Reply by Faarhad** [reply]
Sir, I found the solution.
after running the service successfully, I ran:
`export http_proxy=socks5://127.0.0.1:10808 `
now all traffic passes through v2ray proxy (socs5)
for anyone searching the same problem, for validating that all settings worked correctly run:
`curl http://ipinfo.io/ip`
it should return your proxy IP
good luck guys and thank you @mydogshitgold for your kind replies.

**Reply by mydogshitgold** [reply]
This won't work very well, and you also forget `https_proxy` variable.

If you want to route all traffic from your server, use a transparent proxy; only a handful of programs like curl care about the `http_proxy` / `https_proxy` environment variables, which won't work for a lot of programs. There are also many programs that do not support a socks proxy at all.

**Reply by Faarhad** [reply]
Hi again @mydogshitgold 
You were right, http(s)_proxy is not an operational manner to use proxy in case like "apt-get"
It's working but it's extremely slow. (21kb/sec)
Any help on this would be appreciated sir.

**Reply by Faarhad** [reply]
[This link](https://dongdongbh.tech/blog/tproxy) use a method as you mentioned, transparent proxy.
is it a good practice to follow?
